### PR TITLE
AWS AppSync post

### PR DESCRIPTION
**Pull request by ggotti**
Post on the usage of AWS AppSync and AWS Amplify.
- Description of what AWS AppSync is
- How to use it to generate GraphQL schema
- Sample implementation, called PonyVote

### CODE REVIEW

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+}
+```
+
+The most confusing part of the schema definition is the `schema` special type, which must have a `query` type attribute, and may have a `mutation` type. In additional, AppSync uses a `subscription` type to indicate which data will need to be pushed back to subscribed clients. Once the schema has been successfully loaded (it has inbuild validation), the next step is to "Create Resources". This allows you to automagically create DynamoDB tables based on the schema. It will also extend your GraphQL schema to add common mutations and queries to your schema.
+
+Once a Data Source has been added, you can then attaching custom Resolvers through the schema page. This entire setup reminds me of the old Lambda mapping behaviour, but I'm sure it'll improve over time. The resolver is a [Velocity Template](http://velocity.apache.org/engine/1.7/user-guide.html), which you use to generate a JSON request to the backend service, in this instance, DynamoDB. I can see people just ignoring them and going straight to Lambda as a proxy. To demonstrate the complexity within the template, following request-mapper is used to increment the vote counter for the `vote` mutation.
```
Comment (carvallegro) [inline posts/aws-appsync-is-coming-for-your-api.md]: you can then **attach**

Comment (carvallegro) [inline posts/aws-appsync-is-coming-for-your-api.md]: "I can see people just ignoring them and going straight to Lambda as a proxy." Why? How does that link to the rest of the paragraph?

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+---
```
Comment (carvallegro) [inline posts/aws-appsync-is-coming-for-your-api.md]: You should add some context around your example application sooner that where it is right now. Readers might be surprised when reading "unicorn", "carnival", "vote" and such. Give 'em context!

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+mutation iLikeUnicorns {
+  vote (candidateId: "unicorn") {
+    votes
+  }
+}
+```
+
+One of my favourite parts of the Queries page is that it actually uses the same GrapQL endpoint as your implementation. You can therefore use your browser's inspection tool to see the format of the GraphQL `POST` request if you're having trouble with your client side implementation.
+
+# Developing an Application: PonyVote
```
Comment (carvallegro) [inline posts/aws-appsync-is-coming-for-your-api.md]: You should give pointer about your application. What are the role of each component, why you designed it this way, etc. 
Probably better to do it in the README of https://github.com/ggotti/ponyvote

File: posts/aws-appsync-is-coming-for-your-api.md (hunk 1 above)
Comment (ggotti) [reply posts/aws-appsync-is-coming-for-your-api.md]: fixed 🦄

Comment (ggotti) [reply posts/aws-appsync-is-coming-for-your-api.md]: I've moved this down so it flows better.

File: posts/aws-appsync-is-coming-for-your-api.md (hunk 2 above)
Comment (ggotti) [reply posts/aws-appsync-is-coming-for-your-api.md]: I've modified the text slightly to mention the app.

File: posts/aws-appsync-is-coming-for-your-api.md (hunk 3 above)
Comment (ggotti) [reply posts/aws-appsync-is-coming-for-your-api.md]: I'll add it to the readme

**Comment by ggotti** [comment]
@dsharpe would you mind having a look over this?

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+- The `awsmobile-cli` didn't seem to respect my environment profile settings whilst trying to configure it. Make sure you use `awsmobile configure aws --profile SUPER_PROFILE_NAME` when you call it to explicitly point it at the correct profile.
+
+# Where **would** I use it?
+GraphQL is itself great for situations where you have rapidly changing front-end requirements, and you're application is query heavy. For those cases, I think AppSync is an incredibly powerful tool for quickly developing MVP type applications, and for low-volume applications (and hence, lower running costs). The advantage of using a standardised interface such as GraphQL is that if your application is getting too expensive to run on the platform, you can always move it without needing to re-write your frontend.
+
+# Where **wouldn't** I use it?
+It doesn't make sense to use a service like this for high-volume applications. If you're looking at replacing something high-volume, such as a consumer banking site, you'd be better off using something more bespoke.
```
Comment (dsharpe) [inline posts/aws-appsync-is-coming-for-your-api.md]: Any recommendations? Or afraid to open that can of worms?

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		@aws_subscribe(mutations: ["vote"])
+}
+
+schema {
+	query: Query
+	mutation: Mutation
+	subscription: Subscription
+}
+```
+
```
[the remark this answers is not in the document]
Comment (ggotti) [reply posts/aws-appsync-is-coming-for-your-api.md]: I've added a description about how the subscription mechanism works at a high level.

File: posts/aws-appsync-is-coming-for-your-api.md [hunk 6]
```diff
@@ -113,12 +113,14 @@ To give it a try, I developed a sample app, called [PonyVote](https://github.com
...
 
 # Where **would** I use it?
-GraphQL is itself great for situations where you have rapidly changing front-end requirements, and you're application is query heavy. For those cases, I think AppSync is an incredibly powerful tool for quickly developing MVP type applications, and for low-volume applications (and hence, lower running costs). The advantage of using a standardised interface such as GraphQL is that if your application is getting too expensive to run on the platform, you can always move it without needing to re-write your frontend.
+GraphQL is itself great for situations where you have rapidly changing front-end requirements, and youre application is query heavy. For those cases, I think AppSync is an incredibly powerful tool for quickly developing MVP type applications, and for low-volume applications (and hence, lower running costs). The advantage of using a standardised interface such as GraphQL is that if your application is getting too expensive to run on the platform, you can always move it without needing to re-write your frontend.
```
Comment (dsharpe) [inline posts/aws-appsync-is-coming-for-your-api.md]: "youre" -> "your"

**Comment by ggotti** [comment]
Hey @aseigneurin, I believe this should be good to go.

**Comment by aseigneurin** [comment]
Published: https://blog.ippon.tech/aws-appsync-is-coming-for-your-api/

**Pull request merged**